internal/driver/checkstatuses.go

```diff
@@ -27,40 +32,47 @@ func (d *Driver) checkStatuses() {
-						// already contains them, the credentials will be found (whether they are valid or invalid).
-						device.Protocols[OnvifProtocol][MACAddress] = mac
-					}
-				}
+			d.checkStatusOfDevice(device)
+		}()
+	}
+	wg.Wait()
+}
+
+// checkStatusOfDevice checks the status of an individual device
+func (d *Driver) checkStatusOfDevice(device models.Device) {
+	d.lc.Debugf("checking status of device %s", device.Name)
+
+	// if device is unknown, and missing a MAC Address, try and determine the MAC address via the endpoint reference
+	if strings.HasPrefix(device.Name, UnknownDevicePrefix) && device.Protocols[OnvifProtocol][MACAddress] == "" {
+		if endpointRefAddr := device.Protocols[OnvifProtocol][EndpointRefAddress]; endpointRefAddr != "" {
+			if mac := d.macAddressMapper.MatchEndpointRefAddressToMAC(endpointRefAddr); mac != "" {
+				// the mac address for the device was found, so set it here which will allow the
+				// code below to use the mac address for looking up the credentials. Because the mac mapper
+				// already contains them, the credentials will be found (whether they are valid or invalid).
+				device.Protocols[OnvifProtocol][MACAddress] = mac
 			}
+		}
+	}
+
+	status := d.testConnectionMethods(device)
+	if statusChanged, updateDeviceStatusErr := d.updateDeviceStatus(device.Name, status); updateDeviceStatusErr != nil {
+		d.lc.Warnf("Could not update device status for device %s: %s", device.Name, updateDeviceStatusErr.Error())
 
-			status := d.testConnectionMethods(device)
-			if statusChanged, updateDeviceStatusErr := d.updateDeviceStatus(device.Name, status); updateDeviceStatusErr != nil {
-				d.lc.Warnf("Could not update device status for device %s: %s", device.Name, updateDeviceStatusErr.Error())
-
-			} else if statusChanged && status == UpWithAuth {
-				d.lc.Infof("Device %s is now %s, refreshing the device information.", device.Name, UpWithAuth)
-				go func() { // refresh the device information in the background
-					if refreshErr := d.refreshDevice(device); refreshErr != nil {
-						d.lc.Warnf("An error occurred while refreshing the device %s: %s",
-							device.Name, refreshErr.Error())
-					}
-				}()
+	} else if statusChanged && status == UpWithAuth {
+		d.lc.Infof("Device %s is now %s, refreshing the device information.", device.Name, UpWithAuth)
+		go func() { // refresh the device information in the background
+			if refreshErr := d.refreshDevice(device); refreshErr != nil {
+				d.lc.Warnf("An error occurred while refreshing the device %s: %s",
+					device.Name, refreshErr.Error())
 			}
 		}()
 	}
-	wg.Wait()
+
+	d.lc.Debugf("device %s status is %s", device.Name, status)
 }
 
 // testConnectionMethods will try to determine the state using different device calls
 // and return the most accurate status
-// Higher degrees of connection are tested first, because if they
-// succeed, the lower levels of connection will too
```

Comment:
should we leave this comment in?

Reply:
I will remove it. I think it's removed on my local machine

Reply:
Should it be removed or still kept in? Seems like it should not be removed.

Reply:
I was actually asking if we should _not_ remove it.

Reply:
Right, my bad I had gotten confused on whether I had ultimately removed that line or not.